File: ta_lib/strategies/signal/src/qstick_flip.rs
use base::prelude::*;
use core::prelude::*;
use trend::qstick;

pub struct QSTICKFlipSignal {
    smooth_type: Smooth,
    period: usize,
}

impl QSTICKFlipSignal {
    pub fn new(smooth_type: Smooth, period: f32) -> Self {
        Self {
            smooth_type,
            period: period as usize,
        }
    }
}

impl Signal for QSTICKFlipSignal {
    fn lookback(&self) -> usize {
        self.period
    }

    fn generate(&self, data: &OHLCVSeries) -> (Series<bool>, Series<bool>) {
        let qstick = qstick(&data.open, &data.close, self.smooth_type, self.period);

        (
            qstick.cross_over(&ZERO_LINE),
            qstick.cross_under(&ZERO_LINE),
        )
    }
}

File: strategy/generator/signal/qstick_flip.py
from dataclasses import dataclass

from core.models.parameter import (
    Parameter,
    RandomParameter,
    StaticParameter,
)
from core.models.smooth import Smooth

from .base import Signal, SignalType


@dataclass(frozen=True)
class QstickFlipSignal(Signal):
    type: SignalType = SignalType.QstickFlip
    smooth_type: Parameter = StaticParameter(Smooth.EMA)
    period: Parameter = RandomParameter(10.0, 15.0, 1.0)
